examples/seismic/model.py

```diff
@@ -397,3 +397,105 @@ def critical_dt(self):
         # for further details:
         dt = .85*np.min(self.spacing) / (np.sqrt(self.grid.dim)*self.maxvp)
         return self.dtype("%.3e" % dt)
+
+class ModelViscoacoustic(GenericModel):
+    """
+    The physical model used in seismic inversion processes.
+
+    Parameters
+    ----------
+    origin : tuple of floats
+        Origin of the model in m as a tuple in (x,y,z) order.
+    spacing : tuple of floats
+        Grid size in m as a Tuple in (x,y,z) order.
+    shape : tuple of int
+        Number of grid points size in (x,y,z) order.
+    space_order : int
+        Order of the spatial stencil discretisation.
+    vp : array_like or float
+        Velocity in km/s.
+    rho : float or array, optional
+        Density in kg/cm^3 (rho=1 for water).
+    qp : float or array, optional
+        P-wave quality factor (dimensionless).
+    f0: float or Scalar, optional
+        The dominant frequency
+    nbl : int, optional
+        The number of absorbin layers for boundary damping.
+    dtype : np.float32 or np.float64
+        Defaults to 32.
+
+    The `ModelViscoacoustic` provides two symbolic data objects for the
+    creation of seismic wave propagation operators:
+
+    m : array_like or float
+        The square slowness of the wave.
+    damp : Function
+        The damping field for absorbing boundary condition.
+    """
+    def __init__(self, origin, spacing, shape, space_order, vp, qp, rho, nbl=20, 
+                 subdomains=(), dtype=np.float32):
+        super(ModelViscoacoustic, self).__init__(origin, spacing, shape, space_order,
+                                                nbl=nbl, subdomains=subdomains, 
+                                                dtype=dtype, damp_mask=True)
+
+        self._vp = self._gen_phys_param(vp, 'vp', space_order)
+
+        self.qp = self._gen_phys_param(qp, 'qp', space_order, is_param=True)
+
+        self.rho = self._gen_phys_param(rho, 'rho', space_order, is_param=True)
+
+        self.irho = self._gen_phys_param(1./rho, 'irho', space_order, is_param=True)
+        
+    @property
+    def _max_vp(self):
+        return mmax(self.vp)
+
+    @property
+    def critical_dt(self):
+        """
+        Critical computational time step value from the CFL condition.
+        """
+        # For a fixed time order this number decreases as the space order increases.
+        #
+        # The CFL condtion is then given by
+        # dt <= coeff * h / (max(velocity))
+        coeff = 0.38 if len(self.shape) == 3 else 0.42
+        dt = self.dtype(coeff * np.min(self.spacing) / (self._max_vp))
+        return self.dtype("%.3e" % dt)
```

Comment:
This is the acoustic one, is it still correct for visco-acoustic?

Reply:
Yes. According to the work of Robertsson et al. (1994)

https://library.seg.org/doi/pdf/10.1190/1.1443701,

The stability criteria for the viscoelastic schemes are approximately the same as for the analog elastic schemes. Thus we can consider the viscoacoustic case as a particularity of the viscoelastic case.

However, even if the stability criterion is approximately the same, we intend in the future to add the appropriate approximations for each equation.

Reply:
Perhaps useful to add this as a comment to the code?

Reply:
Yes, it can be useful.

Reply:
>we intend in the future to add the appropriate approximations for each equation.

Not sure what you mean

Reply:
I meant that even though it is reasonable to use the stability condition of the acoustic case for the implemented viscoacoustic equations, I have the objective of calculating the viscoacoustic stability condition for each equation and adding it in the future.

Reply:
Ok that's fine, wouldn't prevent to inherit from Model and only implement dt though